<img src="images/logodwengo.png" alt="Banner" style="width: 150px;"/>

<div style='color: #690027;' markdown="1">
    <h1>FUNCTIES</h1> 
</div>

<div class="alert alert-block alert-success">
Python kent heel wat ingebouwde functies, zoals <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">print()</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">input()</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">type()</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">int()</span> en <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">len()</span>. In deze notebooks leer je zelf functies definiëren. Je kan deze nadien ook gebruiken.
</div>

### Nodige modules importeren

In [1]:
import math         # om vierkantswortels en pi te kunnen gebruiken

<div class="alert alert-block alert-warning"> 
Meer voorbeelden van het gebruik van de module math vind je in de notebook 'Rekenen'.<br>
</div>

<div style='color: #690027;' markdown="1">
    <h2>1. Functie</h2> 
</div>

Op het schema *invoer - verwerking - uitvoer* zie je dat een bepaalde invoer die je aan het programma geeft, een uitvoer kan geven. 
<img src="images/invoerverwerkinguitvoer.png" alt="Banner" style="width: 600px;"/><br>
Soms krijg je echter geen uitvoer maar een foutboodschap bv. Je kan in elk geval hoogstens 1 uitvoer verkijgen, die weliswaar meerdere elementen kan bevatten. 

<div style='color: #690027;' markdown="1">
    <h3>1.1 Verwerking</h3> 
</div>

### Verdubbelen

<img src="images/functie1.png" alt="Banner" style="width: 300px;"/><br>

<table>
    <thead>
        <tr>
            <th><p align="center">Invoer</th>  
            <th><p align="center">Verwerking</th>
            <th>&nbsp;</th>
            <th><p align="center">Uitvoer</th>    
    </thead>
    <tr> <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2 . 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;6  
    <tr> <td> <p align="center">&nbsp;&nbsp;-10        <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2 . (-10)<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;-20    
    <tr> <td>  <p align="center">&nbsp;&nbsp;2,5       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2 . 2,5<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;5       
            <tr> <td>  <p align="center">&nbsp;&nbsp;x       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2 . x<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2x    
</table>

<center> In wiskunde noteert men: $f(x) = x$ </center>

### Verminderen met 3

<img src="images/functie3.png" alt="Banner" style="width: 300px;"/><br>

<table>
    <thead>
        <tr>
            <th><p align="center">Invoer</th>  
            <th><p align="center">Verwerking</th>
            <th>&nbsp;</th>
            <th><p align="center">Uitvoer</th>    
    </thead>
    <tr> <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3 -3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;0  
    <tr> <td> <p align="center">&nbsp;&nbsp;-10        <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;-10 -3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;-13    
    <tr> <td>  <p align="center">&nbsp;&nbsp;2,5       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2,5 - 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;-0,5       
    <tr> <td>  <p align="center">&nbsp;&nbsp;x       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;x - 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;x-3        
</table>

<center> In wiskunde noteert men: $f(x) = x -3$ </center>

### 20 % nemen

<img src="images/functie2.png" alt="Banner" style="width: 300px;"/><br>

<table>
    <thead>
        <tr>
            <th><p align="center">Invoer</th>  
            <th><p align="center">Verwerking</th>
            <th>&nbsp;</th>
            <th><p align="center">Uitvoer</th>    
    </thead>
    <tr> <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3 . 0,20<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;0,6  
    <tr> <td> <p align="center">&nbsp;&nbsp;10        <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;10 . 0,20<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2    
    <tr> <td>  <p align="center">&nbsp;&nbsp;2,5       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2,5 . 0,20<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;0,5       
    <tr> <td>  <p align="center">&nbsp;&nbsp;x       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;x . 0,20<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;0,2x         
</table>

<center> In wiskunde noteert men: $f(x) = 0,2 \; x$ </center>

### Rest bij gehele deling door 3

<img src="images/functie4.png" alt="Banner" style="width: 300px;"/><br>

<table>
    <thead>
        <tr>
            <th><p align="center">Invoer</th>  
            <th><p align="center">Verwerking</th>
            <th>&nbsp;</th>
            <th><p align="center">Uitvoer</th>    
    </thead>
    <tr> <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;3 mod 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;0  
    <tr> <td> <p align="center">&nbsp;&nbsp;10        <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;10 mod 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;1   
    <tr> <td>  <p align="center">&nbsp;&nbsp;26       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;26 mod 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2     
    <tr> <td>  <p align="center">&nbsp;&nbsp;x       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;x mod 3<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;x mod 3 
</table>

<center> In wiskunde noteert men: $f(x) = x \; mod \; 3$ </center>

### Vierkantswortel nemen

<img src="images/functie5.png" alt="Banner" style="width: 300px;"/><br>

<table>
    <thead>
        <tr>
            <th><p align="center">Invoer</th>  
            <th><p align="center">Verwerking</th>
            <th>&nbsp;</th>
            <th><p align="center">Uitvoer</th>    
    </thead>
    <tr> <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;4          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;$\sqrt4$<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;2  
    <tr> <td> <p align="center">&nbsp;&nbsp;-4        <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;/<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;/  
    <tr> <td>  <p align="center">&nbsp;&nbsp;100       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;$\sqrt100$<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;10       
    <tr> <td>  <p align="center">&nbsp;&nbsp;x       <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;$\sqrt x$<td> <td> 
        <p align="center">&nbsp;&nbsp;&nbsp;&nbsp;$\sqrt x$
</table>

<center> In wiskunde noteert men: $f(x) = \sqrt x$ </center>

Dit waren alvast enkele voorbeelden van functies.<br> **Elke x-waarde heeft door een functie hoogstens 1 beeld.** <br>
Het **domein** van een functie is de verzameling x'n die effectief een beeld hebben, kan alle reële getallen bevatten, maar ook slechts een deel ervan (zoals bij de vierkantsworteltrekking).<br>
Het **beeld** van een functie is de verzameling van de y-waarden die effectief bereikt worden.

<div class="alert alert-block alert-info">
Als je bepaalde code meerdere keren wilt gebruiken, loont het de moeite om zelf een functie te definiëren.<br> Functies worden gedeclareerd met het sleutelwoord <b>def</b>. Via het sleutelwoord <b>return</b> geven ze een resultaat terug.<br>
Een <b>docstring</b> verduidelijkt het doel van de functie.
</div>

<div style='color: #690027;' markdown="1">
    <h3>1.2 Functies definiëren in Python</h3> 
</div>

Hier krijg je voorbeelden van hoe je zelf een functie definieert in Python.

### Verdubbelen

Voer de volgende drie code-cellen uit. Begrijp je wat er gebeurt?

In [2]:
def verdubbelen(x):
    """Verdubbelen."""
    dubbel = 2 * x
    return dubbel

In [3]:
verdubbelen(10)

20

In [4]:
verdubbelen(math.pi)

6.283185307179586

<div class="alert alert-box alert-success">
    Let bij het opstellen van de definitie op de <b>dubbelepunt</b> en de <b>indentatie</b> (het inspringen met vier karakters). De functie neemt een parameter x als <b>invoer</b> en geeft het dubbele van x terug als <b>uitvoer</b>. <br>
    De tekst tussen de aanhalingstekens is een <b>docstring</b> en dient om te verduidelijken waarvoor de functie dient. Een docstring begint met een hoofdletter en eindigt met een punt.
</div>

### Verminderen met 3

In [5]:
def verminderen(x):
    """Verminderen met 3."""
    resultaat = x - 3
    return resultaat

In [6]:
verminderen(3)

0

In [7]:
verminderen(2.5)

-0.5

Voor de voorbeelden met de rest en de vierkantswortel is het niet zinvol om er een functie voor te schrijven, aangezien beide al in Python bestaan als voorgedefinieerde functie, nl. `%` en `sqrt()`.<br>
Hieronder vind je wat voorbeelden.

In [8]:
math.sqrt(100)

10.0

In [9]:
math.sqrt(-4)

ValueError: math domain error

In [10]:
26 % 3

2

In [11]:
65 % 5

0

### Opdracht 1.1: 20% nemen

Implementeer de functie `procentnemen()` voor het resterende voorbeeld van hierboven: 20 % nemen. Test de code uit op enkele getalvoorbeelden.

In [12]:
# voorbeeldscript
def procentnemen(x):
    """20 procent nemen."""
    percentage = x * 0.20
    return percentage 

In [13]:
# voorbeeldscript
procentnemen(10)

2.0

In [14]:
# voorbeeldscript
procentnemen(3)

0.6000000000000001


<div class="alert alert-block alert-info">
<img src="images/patroonherkenningdwengo.png" alt="Banner" style="width: 150px;"/><br>
De definitie van een functie bestaat altijd uit vier delen:<ol>
        <li>benaming: naam van de functie</li>
        <li>docstring: beschrijving van de functie</li>
        <li>definitie: script dat alle bewerkingen uitvoert</li>
        <li>return: output van de functie</li></ol>
<img src="images/decompositiedwengo.png" alt="Banner" style="width: 150px;"/><br>
    Gebruikmaken van functies is een vorm van <b>abstractie</b>. Je brengt immers verschillende stappen in een algoritme samen en gebruikt die stappen dan ineens via de functie. Zo worden sommige details van het algoritme verborgen.
<img src="images/abstractiedwengo.png" alt="Banner" style="width: 150px;"/>
<img src="images/algoritmedwengo.png" alt="Banner" style="width: 150px;"/>    
</div>

<div style='color: #690027;' markdown="1">
    <h2>2. Functies voor het berekenen omtrek, oppervlakte en inhoud</h2> 
</div>

<div style='color: #690027;' markdown="1">
    <h3>2.1 Voorbeelden</h3> 
</div>

### Omtrek vierkant

De omtrek van een vierkant met zijde $z$ is vier keer de zijde, dus $4z$. De omtrek wordt dus bepaald door de afmeting van de zijde. $z$ wordt daarom meegegeven als parameter aan de functie `omtrekvierkant()`. 

In [15]:
def omtrekvierkant(z):
    """Omtrek vierkant met zijde z."""
    omtrek = 4 * z
    return omtrek

In [16]:
# toepassen
# vierkant met zijde 6
omtrekvierkant(6)

24

### Oppervlakte rechthoek

De oppervlakte van een rechthoek met breedte $b$ en lengte $l$ is het product van $l$ en $b$. Hier worden dus twee parameters meegegeven aan de functie `oppervlakterechthoek()`.  

In [17]:
def oppervlakterechthoek(l, b):
    """Oppervlakte rechthoek met lengte l en breedte b."""
    oppervlakte = l * b
    return oppervlakte

In [18]:
# toepassen
# rechthoek met lengte 7 en breedte 3
oppervlakterechthoek(7, 3)

21

### Inhoud van een piramide of kegel gebruikmakend van grondvlak en hoogte.

De inhoud van een piramide of kegel is een derde van het product van de oppervlakte van het grondvlak en de hoogte. Definieer hiervoor een functie. 

In [19]:
def inhoudpir(s, h):
    """Inhoud piramide of kegel met oppervlakte grondvlak s en hoogte h."""
    inhoud = s * h /3
    return inhoud

Dit betekent dat je om deze formule te kunnen gebruiken, je de oppervlakte van het grondvlak moet kennen. Di kan je eventueel met een andere functie berekenen. 

<div class="alert alert-block alert-warning"> 
    Bekijk eventueel de notebook 'Rekenen' om te herhalen hoe je rekenkundige bewerkingen ingeeft in Python.
</div>

<div style='color: #690027;' markdown="1">
    <h3>2.2 Oefeningen</h3> 
</div>

### Oefening 2.1

Stel functies op voor het berekenen van:
- de omtrek van een rechthoek;
- de oppervlakte van een vierkant;
- de oppervlakte van een cirkel;
- de oppervlakte van een driehoek;
- de inhoud van een balk;
- de som van de lengtes van alle ribben van een balk.

In [20]:
# voorbeeldscript 1
def omtrekrechthoek(l, b):
    """Omtrek rechthoek met breedte b en lengte l."""
    omtrek = 2 * (l + b)
    return omtrek

In [21]:
# voorbeeldscript 2
def oppvierkant(z):
    """Oppervlakte vierkant met zijde z."""
    opp = z**2
    return opp

In [22]:
# voorbeeldscript 3
def oppcirkel(r):
    """Oppervlakte cirkel met straal r."""
    opp = math.pi * r**2
    return opp

In [23]:
# voorbeeldscript 4
def oppdriehoek(b, h):
    """Oppervlakte driehoek met basis b en hoogte h."""
    opp = b * h / 2
    return opp

In [38]:
# voorbeeldscript 5
def inhoudbalk(s, h):
    """Inhoud balk met oppervlakte grondvlak s en hoogte h."""
    inh = s * h 
    return inh

In [39]:
# voorbeeldscript 5 alternatief
def inhoudbalk2(l, d, h):
    """Inhoud balk met lengte l, diepte d en hoogte h."""
    inh = l * d * h 
    return inh

In [26]:
# voorbeeldscript 6
def lengteribbenbalk(l, d, h):
    """Lengte ribben balk met diepte d, lengte l en hoogte h."""
    lengte = 4 * (l + d + h)
    return lengte

### Oefening 2.2

Gebruik de functies om de antwoorden op de volgende oefeningen te berekenen. Hier en daar zul je nog extra code moeten schrijven.

- De oppervlakte van een vierkant met zijde 4 cm.

In [27]:
oppvierkant(4)

16

Antwoord:

Antwoord: Het vierkant heeft een oppervlakte van 16 cm^2.

- De omtrek van een rechthoek van 3 cm op 7 cm.

In [28]:
omtrekrechthoek(3, 7)

20

Antwoord:

Antwoord: De omtrek is 20 cm.

- De oppervlakte van een cirkel met diameter 11 dm.

In [29]:
d = 11
r = d /2
oppcirkel(r)

95.03317777109125

Antwoord:

Antwoord: De oppervlakte is ongeveer 95 dm^2.

- Bereken de oppervlakte van een cirkel met omtrek 8 meter.

In [30]:
omtrek = 8
r = omtrek /(2 * math.pi)
oppcirkel(r)

5.092958178940651

Antwoord:

Antwoord: De oppervlakte is 5 m^2.

- Bereken de oppervlakte van een driehoek met basis 4 dm en hoogte 100 cm. 

In [31]:
oppdriehoek(4, 10)

20.0

Antwoord:

Antwoord: De oppervlakte is 20 dm^2

- Bereken de inhoud van een kegel met een cirkel met straal 10 cm als grondvlak en hoogte 12 cm.

In [33]:
s = oppcirkel(10)
inhoudpir(s, 12)

1256.6370614359173

Antwoord:

Antwoord: De kegel heeft een inhoud van 1,3 dm^3

- Bereken de inhoud van de cilinder die je bekomt door een vierkant blad met zijde 2 dm toe te rollen (m.a.w. linkerzijde en rechterzijde aan elkaar kleven, overlap van 1 cm). Schrijf de nodige functies als die ontbreken. 

In [40]:
omtrekgrondvlak = 20 - 1
hoogte = 20
r = omtrekgrondvlak / (2 * math.pi)
s = oppcirkel(r)
inhoudbalk(s, hoogte)

574.5493445617421

Antwoord:

Antwoord: De cilinder heeft een inhoud van 575 cm^3.

- Aan de ingang van het Louvre in Parijs staat een glazen piramide met vierkant grondvlak, en met een hoogte van 21,65 meter en een lengte van 35,42 meter. Bereken de inhoud van deze piramide.

In [41]:
z = 35.42
h = 21.65
s = oppvierkant(z)
inhoudpir(s, h)

9053.859686666667

Antwoord:

Antwoord: De inhoud bedraagt ongeveeer 9054 m^3.

### Oefening 2.3

Schrijf een Python-script dat de inhoud van een balk zal berekenen. Het script vraagt de gebruiker wat de gewenste afmetingen van de balk zijn. Na afloop wordt er een zin afgedrukt waarin de inhoud aan de gebruiker wordt verteld.

In [42]:
# voorbeeldscript
lengte = float(input("Lengte balk in cm : "))
diepte = float(input("diepte balk in cm: "))
hoogte = float(input("hoogte balk in cm: "))

inh = inhoudbalk2(lengte, diepte, hoogte)

print("De inhoud van de balk is", inh, "cm^3.")

Lengte balk in cm : 17
diepte balk in cm: 5
hoogte balk in cm: 3
De inhoud van de balk is 255.0 cm^3.


<div class="alert alert-block alert-warning"> 
Uitleg over input en output vind je terug in de notebook "InputOutput".
</div>

<div style='color: #690027;' markdown="1">
    <h2>3. Sleutelwoorden</h2> 
</div>

<div class="alert alert-block alert-info">
<span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">def</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">else</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">False</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">for</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">if</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">import</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">in</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">return</span>, <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">while</span> zijn *sleutelwoorden* in Python. In een script ingevoerd in Jupyter notebook kan je ze herkennen omdat ze vetgedrukt zijn. <br><b>Sleutelwoorden zijn woorden die in Python gereserveerd zijn voor een bepaald doel.</b> Kies dus nooit een sleutelwoord als naam voor een variabele. 
De sleutelwoorden zijn hoofdlettergevoelig en moeten exact geschreven worden zoals ze zijn gereserveerd. 

Twijfel je of een bepaald woord een sleutelwoord is? Dan kan je via het volgende Python-script alle sleutelwoorden opvragen. 
</div>

In [43]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

<div class="alert alert-block alert-warning"> 
Kader 2: Verwijzing
</div>

<div class="alert alert-block alert-danger"> 
Kader 4: Let op
</div>

<img src="images/cclic.png" alt="Banner" align="left" style="width:100px;"/><br><br>
Notebook Python in wiskunde, zie Computationeel denken - Programmeren in Python van <a href="http://www.aiopschool.be">AI Op School</a>, van F. wyffels, B. Van de Velde & N. Gesquière is in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>. 